MidasFrontend/ArduinoHelper.cxx

```diff
@@ -0,0 +1,195 @@
+    return false;
+}
+
+static bool attempt_move_axis(AxisId axis, int32_t cur_pos_counts, float dest_mm, float vel_mm_s)
+{
+    // Target position
+    int32_t target_counts = mm_to_cts(dest_mm);
+
+    // Relative displacement
+    int32_t disp_counts = (target_counts - cur_pos_counts);
+
+    // Direction
+    AxisDirection dir = get_direction(disp_counts);
+
+    // Velocity
+    uint32_t vel_steps_s = mm_to_steps(vel_mm_s);
+
+    // Send command
+    SerialResult ser_res;
+    AxisResult axis_res;
+    ser_res = comm.move(axis, dir, vel_steps_s, abs(disp_counts), &axis_res, MSG_RECEIVE_TIMEOUT);
+
+    // Handle results
+    return (handle_serial_result(ser_res) && handle_axis_result(axis_res));
+}
+
+/*****************************************************************************/
+/*                             PUBLIC FUNCTIONS                              */
+/*****************************************************************************/
+
+/**
+ * @brief Establishes a serial connection to the Arduino
+ * 
+ * This includes opening the serial device as well as waiting to receive a
+ * ping message from the Arduino to validate that it is running
+ * 
+ * @return true if the connection was successfully established, otherwise false
+ */
+bool arduino_connect(char *device_file)
+{
+    // Open the serial device
+    device.set_device_file(device_file);
+    if (!device.ser_connect(BAUD_RATE)) return false;
+    device.ser_flush();
+    
+    printf("Waiting for Arduino...");
+    while (!(comm.check_for_message() && comm.received_message().id == MSG_ID_PING)) {
+        // Wait for ping
+    }
+    // There might be more ping messages sitting in the buffer, so flush them all out
+    device.ser_flush();
+    printf("Connected!\n");
+
+    return true;
+}
+
+/**
+ * @brief Disconnects from the Arduino
+ */
+void arduino_disconnect()
+{
+    device.ser_disconnect();
+}
+
+/**
+ * @brief Attempts to command the Arduino to move to the provided destination at the
+ *        provided velocity
+ * 
+ * @param dest_mm   Pointer to two floats (the absolute x and y coordinates in mm)
+ * @param vel_mm_s  Pointer to two floats (the x and y velocities in mm/s)
+ */
+void arduino_attempt_move(float *dest_mm, float *vel_mm_s)
+{
+    if (!validate_move_params(dest_mm, vel_mm_s)) return;
+
+    // Get current position
+    PositionMsgData cur_pos_counts;
+    if (!handle_serial_result(comm.get_position(&cur_pos_counts, MSG_RECEIVE_TIMEOUT))) return;
+
+    // Attempt movement
+    bool x_success = attempt_move_axis(AXIS_X, cur_pos_counts.x_counts, dest_mm[AXIS_X], vel_mm_s[AXIS_X]);
+    bool y_success = attempt_move_axis(AXIS_Y, cur_pos_counts.y_counts, dest_mm[AXIS_Y], vel_mm_s[AXIS_Y]);
+
+    // Handle results
+    if (x_success && y_success) {
+        printf("Moving to position (%f mm, %f mm)\n", dest_mm[AXIS_X], dest_mm[AXIS_Y]);
+        printf("Moving with velocity (%f mm/s, %f mm/s)\n", vel_mm_s[AXIS_X], vel_mm_s[AXIS_Y]);
+    }
```

Comment:
Should we have a case where we display if x succeeded but y didn't, and vice versa?

Reply:
I'm saying if either of them failed then the "move" as a whole failed, and we stop both axes. I don't think we want to do half moves where one axis moves, but the other doesn't. Still working on nicer error reporting.

Reply:
makes sense :)